### PR TITLE
[Storage] Add structured message CRC64 content validation for azfile

### PR DESCRIPTION
**Pull request by tanyasethi-msft**
## Summary
- Adds structured message (XSM/1.0) reader/writer infrastructure to azfile and azdatalake internal packages
- Integrates CRC64 content validation into azfile upload (UploadRange) and download (DownloadStream) paths
- Adds `TransferValidationTypeComputeStructuredMessageCRC64` option for azfile clients
- For uploads: wraps request body with SMEncoder, sets `x-ms-structured-body` header
- For downloads: detects structured body response header and wraps body with SMDecoder for transparent CRC64 validation

## Test plan
- [x] Unit tests for structured message encode/decode 
- [ ] Integration tests for upload/download with SM CRC64 (in follow-up PR)

### CODE REVIEW

File: sdk/storage/azfile/internal/shared/structured_message.go [hunk 1]
```diff
@@ -0,0 +1,801 @@
...
+					d.prepareFrame(SMSegmentFooterSize)
+					d.state = decStateSegFooter
+				} else {
+					d.advanceAfterSegment()
+				}
+			}
+			if err != nil && d.segRemain > 0 {
+				d.setError(fmt.Errorf("segment %d: %s", d.segIndex, err))
+				return totalOut, d.err
+			}
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azfile/internal/shared/structured_message.go]: Good catch — changed to %w so net.Error is preserved through errors.Is/As for the retry path. Fixed in the latest push.

File: sdk/storage/azfile/internal/shared/structured_message.go [hunk 2]
```diff
@@ -0,0 +1,801 @@
...
+
+func (d *SMDecoder) validateTrailerCRC() error {
+	expected := binary.LittleEndian.Uint64(d.frameBuf[:SMMessageTrailerSize])
+	actual := d.msgCRC.Sum64()
+	if expected != actual {
+		return fmt.Errorf("message trailer CRC64 mismatch (expected 0x%016x, got 0x%016x)", expected, actual)
+	}
+	d.state = decStateDone
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azfile/internal/shared/structured_message.go:774]: Added msgLen validation in validateTrailerCRC — now checks bytesRead against the declared message length before completing. This aligns the streaming decoder with SMDecode behavior.

File: sdk/storage/azdatalake/internal/shared/structured_message.go [hunk 3]
```diff
@@ -0,0 +1,801 @@
...
+
+func (d *SMDecoder) validateTrailerCRC() error {
+	expected := binary.LittleEndian.Uint64(d.frameBuf[:SMMessageTrailerSize])
+	actual := d.msgCRC.Sum64()
+	if expected != actual {
+		return fmt.Errorf("message trailer CRC64 mismatch (expected 0x%016x, got 0x%016x)", expected, actual)
+	}
+	d.state = decStateDone
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azdatalake/internal/shared/structured_message.go]: Same fix applied to azdatalake copy.

File: sdk/storage/azfile/internal/shared/structured_message.go [hunk 4]
```diff
@@ -0,0 +1,801 @@
...
+	}
+
+	totalDataLen := len(data)
+	numSegments := totalDataLen / segmentSize
+	if totalDataLen%segmentSize != 0 {
+		numSegments++
+	}
+	if numSegments == 0 {
+		numSegments = 1
+	}
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azfile/internal/shared/structured_message.go:54]: Added overflow protection to SMEncode matching what NewSMEncoder already does — auto-increases segment size when numSegments exceeds uint16 max.

File: sdk/storage/azdatalake/internal/shared/structured_message.go [hunk 5]
```diff
@@ -0,0 +1,801 @@
...
+	}
+
+	totalDataLen := len(data)
+	numSegments := totalDataLen / segmentSize
+	if totalDataLen%segmentSize != 0 {
+		numSegments++
+	}
+	if numSegments == 0 {
+		numSegments = 1
+	}
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azdatalake/internal/shared/structured_message.go]: Same fix applied to azdatalake copy.

File: sdk/storage/azfile/file/client.go [hunk 6]
```diff
@@ -260,6 +262,19 @@ func (f *Client) UploadRange(ctx context.Context, offset int64, body io.ReadSeek
 		return UploadRangeResponse{}, err
 	}
 
+	if options != nil && options.TransactionalValidation != nil {
+		if _, ok := options.TransactionalValidation.(TransferValidationTypeMD5); !ok {
+			body, err = options.TransactionalValidation.Apply(body, uploadRangeOptions)
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azfile/file/client.go:266]: Integration tests are in a follow-up PR (#27264). Will consider adding pipeline/mock transport tests for request header verification as well.

File: sdk/storage/azfile/file/client.go [hunk 7]
```diff
@@ -579,11 +618,16 @@ func (f *Client) DownloadStream(ctx context.Context, options *DownloadStreamOpti
 		return DownloadStreamResponse{}, err
 	}
 
+	if resp.StructuredBodyType != nil && *resp.StructuredBodyType != "" {
+		resp.Body = shared.NewSMDecoder(resp.Body)
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azfile/file/client.go:624]: Same as above — integration tests in #27264. Mock transport tests for the download wrapping/retry path are a good idea, will add.

File: sdk/storage/azfile/file/constants.go [hunk 8]
```diff
@@ -143,6 +143,22 @@ type TransferValidationType = exported.TransferValidationType
...
+// a CRC64 should be computed during transfer.
+func TransferValidationTypeComputeCRC64() TransferValidationType {
+	return exported.TransferValidationTypeComputeCRC64()
+}
+
+// TransferValidationTypeComputeStructuredMessageCRC64 is a TransferValidationType that computes
+// per-segment CRC64 checksums using the structured message binary format (XSM/1.0).
+// segmentSize specifies the maximum segment size in bytes. Values <= 0 use the default (4 MB).
+func TransferValidationTypeComputeStructuredMessageCRC64(segmentSize int) TransferValidationType {
+	return exported.TransferValidationTypeComputeStructuredMessageCRC64(segmentSize)
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azfile/file/constants.go:150]: Acknowledged — will add CHANGELOG entry before the release.

File: sdk/storage/azfile/file/client.go [hunk 9]
```diff
@@ -260,6 +262,19 @@ func (f *Client) UploadRange(ctx context.Context, offset int64, body io.ReadSeek
...
+	if options != nil && options.TransactionalValidation != nil {
+		if _, ok := options.TransactionalValidation.(TransferValidationTypeMD5); !ok {
+			body, err = options.TransactionalValidation.Apply(body, uploadRangeOptions)
+			if err != nil {
+				return UploadRangeResponse{}, err
+			}
+			contentLength, err = shared.ValidateSeekableStreamAt0AndGetCount(body)
+			if err != nil {
+				return UploadRangeResponse{}, err
+			}
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azfile/file/client.go:279]: Fair point about the azfile upload path. The body assignment in UploadRange already uses the return from Apply, so the encoded body does get sent. But will double-check the Optionalbody vs body flow to make sure Content-Length is consistent.

File: sdk/storage/azfile/file/client.go [hunk 10]
```diff
@@ -579,11 +618,16 @@ func (f *Client) DownloadStream(ctx context.Context, options *DownloadStreamOpti
 		return DownloadStreamResponse{}, err
 	}
 
+	if resp.StructuredBodyType != nil && *resp.StructuredBodyType != "" {
+		resp.Body = shared.NewSMDecoder(resp.Body)
+	}
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azfile/file/client.go:625]: Noted — this is a valid concern for chunked downloads. Will investigate the RetryReader EOF behavior with the SM decoder trailer.

File: sdk/storage/azfile/internal/exported/transfer_validation_option.go [hunk 11]
```diff
@@ -14,6 +20,31 @@ type TransferValidationType interface {
...
 }
 
+// TransferValidationTypeCRC64 is a TransferValidationType used to provide a precomputed CRC64.
+type TransferValidationTypeCRC64 uint64
+
+func (c TransferValidationTypeCRC64) Apply(rsc io.ReadSeekCloser, cfg generated.TransactionalContentSetter) (io.ReadSeekCloser, error) {
+	buf := make([]byte, 8)
+	binary.LittleEndian.PutUint64(buf, uint64(c))
+	cfg.SetCRC64(buf)
+	return rsc, nil
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azfile/internal/exported/transfer_validation_option.go]: The CRC64 types for azfile are exported for consistency with azblob's public API surface. Azure Files does support transactional CRC via structured message (that's what this PR adds). The SetCRC64 no-op is for the header-based transactional CRC which Files doesn't support — we could add a guard there, but structured message CRC64 is the primary path.

File: sdk/storage/azfile/internal/shared/structured_message.go [hunk 12]
```diff
@@ -0,0 +1,801 @@
...
+	buf := d.frameBuf[:SMHeaderSize]
+	d.version = buf[0]
+	if d.version != SMVersion {
+		return fmt.Errorf("unsupported structured message version: %d (expected %d)", d.version, SMVersion)
+	}
+
+	d.msgLen = binary.LittleEndian.Uint64(buf[1:9])
+	d.flags = binary.LittleEndian.Uint16(buf[9:11])
+	d.numSegments = binary.LittleEndian.Uint16(buf[11:13])
+	d.hasCRC = d.flags&SMFlagCRC64 != 0
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azfile/internal/shared/structured_message.go:704]: We specifically want to accept the flags as-is from the service response. The server negotiation via x-ms-structured-body header already guarantees CRC64 will be present. Rejecting on missing flag would break forward compatibility if the service ever sends messages with additional properties.

File: sdk/storage/azfile/internal/shared/structured_message.go [hunk 13]
```diff
@@ -0,0 +1,801 @@
...
+
+	offset := SMHeaderSize
+	var result []byte
+	var msgHasher hash64
+	if hasCRC {
+		msgHasher = crc64.New(CRC64Table)
+	}
+
+	for i := uint16(1); i <= numSegments; i++ {
+		segment, err := decodeSMSegment(smData, offset, i, hasCRC)
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azfile/internal/shared/structured_message.go]: Good catch on the uint16 overflow in the loop counter. Will use a wider counter.

File: sdk/storage/azdatalake/internal/shared/structured_message.go [hunk 14]
```diff
@@ -0,0 +1,801 @@
...
+	buf := d.frameBuf[:SMHeaderSize]
+	d.version = buf[0]
+	if d.version != SMVersion {
+		return fmt.Errorf("unsupported structured message version: %d (expected %d)", d.version, SMVersion)
+	}
+
+	d.msgLen = binary.LittleEndian.Uint64(buf[1:9])
+	d.flags = binary.LittleEndian.Uint16(buf[9:11])
+	d.numSegments = binary.LittleEndian.Uint16(buf[11:13])
+	d.hasCRC = d.flags&SMFlagCRC64 != 0
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azdatalake/internal/shared/structured_message.go]: Same rationale as above — we accept the flags from the server response as-is.

File: sdk/storage/azdatalake/internal/shared/structured_message.go [hunk 15]
```diff
@@ -0,0 +1,801 @@
...
+
+	offset := SMHeaderSize
+	var result []byte
+	var msgHasher hash64
+	if hasCRC {
+		msgHasher = crc64.New(CRC64Table)
+	}
+
+	for i := uint16(1); i <= numSegments; i++ {
+		segment, err := decodeSMSegment(smData, offset, i, hasCRC)
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azdatalake/internal/shared/structured_message.go]: Will fix the loop counter to use a wider type, same as the azfile copy.

File: sdk/storage/azfile/file/models.go [hunk 16]
```diff
@@ -587,6 +593,11 @@ func (o *DownloadStreamOptions) format(fileRequestIntent *generated.ShareTokenIn
 	if o.LeaseAccessConditions != nil {
 		opts.LeaseID = o.LeaseAccessConditions.LeaseID
 	}
+	if o.TransactionalValidation != nil {
+		if h := exported.GetStructuredBodyType(o.TransactionalValidation); h != "" {
+			opts.StructuredBodyType = &h
+		}
+	}
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azfile/file/models.go:600]: This follows the same pattern as azblob. The download options accept TransferValidationType for structured message CRC64 specifically. Other validation types are not applicable for downloads and are silently ignored, matching the blob behavior. We could add explicit rejection for non-SM types if preferred.

File: sdk/storage/azdatalake/internal/shared/structured_message.go
Comment (jhendrixMSFT) [inline sdk/storage/azdatalake/internal/shared/structured_message.go:1]: I believe these two `azdatalake` files were added in error?  I see them on the other PR.

File: sdk/storage/azfile/internal/generated/models.go [hunk 17]
```diff
@@ -14,13 +14,23 @@ import (
...
+	SetCRC64([]byte)
+	SetStructuredBody(bodyType string, contentLength int64)
 }
 
 func (f *FileClientUploadRangeOptions) SetMD5(v []byte) {
 	f.ContentMD5 = v
 }
 
+func (f *FileClientUploadRangeOptions) SetCRC64([]byte) {
+	// no-op: Azure Files does not yet support transactional CRC64 headers on UploadRange
```
Comment (jhendrixMSFT) [inline sdk/storage/azfile/internal/generated/models.go:26]: I'm a bit confused here.  I get that we need this no-op method to satisfy the interface.  However, if this isn't supported, then why is `TransferValidationTypeCRC64` exported in `constants.go`?

File: sdk/storage/azfile/internal/shared/structured_message.go [hunk 18]
```diff
@@ -0,0 +1,812 @@
...
+				// Segment data done
+				if d.hasCRC {
+					d.prepareFrame(SMSegmentFooterSize)
+					d.state = decStateSegFooter
+				} else {
+					d.advanceAfterSegment()
+				}
+			}
+			if err != nil && d.segRemain > 0 {
+				d.setError(fmt.Errorf("segment %d: %w", d.segIndex, err))
```
Comment (jhendrixMSFT) [inline sdk/storage/azfile/internal/shared/structured_message.go]:
```go
fmt.Errorf("segment %d: %s", d.segIndex, err.Error())
```

File: sdk/storage/azfile/internal/generated/models.go (hunk 17 above)
Comment (jhendrixMSFT) [reply sdk/storage/azfile/internal/generated/models.go:26]: Question also applies to `func TransferValidationTypeComputeCRC64()`.

File: sdk/storage/azdatalake/internal/shared/structured_message.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azdatalake/internal/shared/structured_message.go]: Fixed — used uint32 loop counter that casts to uint16 only when passing to decodeSMSegment. Same fix in both azfile and azdatalake.

File: sdk/storage/azdatalake/internal/shared/structured_message.go
Comment (tanyasethi-msft) [reply sdk/storage/azdatalake/internal/shared/structured_message.go:1]: You're right — moved the azdatalake SM files to PR #27263 where they belong. This PR is now azfile-only.

File: sdk/storage/azfile/internal/generated/models.go (hunk 17 above)
Comment (tanyasethi-msft) [reply sdk/storage/azfile/internal/generated/models.go:26]: Good point — removed both TransferValidationTypeCRC64 and TransferValidationTypeComputeCRC64 from azfile's public exports since Azure Files doesn't support the transactional CRC64 header. Only TransferValidationTypeComputeStructuredMessageCRC64 is exported now.

File: sdk/storage/azfile/internal/shared/structured_message.go (hunk 18 above)
Comment (tanyasethi-msft) [reply sdk/storage/azfile/internal/shared/structured_message.go]: We actually changed this from %s to %w intentionally — using %w preserves the original error type through errors.Is/errors.As, which is important here because RetryReader checks for net.Error to decide if a failure is retryable. With %s or .Error(), the concrete type gets erased and transient network errors during segment reads would stop being retried. That was actually one of the Copilot findings we fixed in this round.

File: sdk/storage/azfile/internal/shared/structured_message.go [hunk 19]
```diff
@@ -0,0 +1,812 @@
...
+	}
+
+	msgLen := binary.LittleEndian.Uint64(smData[1:9])
+	if uint64(len(smData)) != msgLen {
+		return SMDecodeResult{}, fmt.Errorf("structured message length mismatch: header says %d, got %d bytes", msgLen, len(smData))
+	}
+
+	flags := binary.LittleEndian.Uint16(smData[9:11])
+	numSegments := binary.LittleEndian.Uint16(smData[11:13])
+	hasCRC := flags&SMFlagCRC64 != 0
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azfile/internal/shared/structured_message.go:203]: The SMDecode function is a general-purpose structured message parser. CRC64 validation is enforced at the transport layer — we only install the decoder when the response header says properties=crc64. A corrupted flags field would also cause the msgLen check to fail since the byte counts wouldn't add up without footer/trailer bytes. Hardcoding a CRC64 requirement here would break forward compatibility if the format gains new property types.

File: sdk/storage/azdatalake/internal/shared/structured_message.go [hunk 20]
```diff
@@ -0,0 +1,812 @@
...
+	}
+
+	msgLen := binary.LittleEndian.Uint64(smData[1:9])
+	if uint64(len(smData)) != msgLen {
+		return SMDecodeResult{}, fmt.Errorf("structured message length mismatch: header says %d, got %d bytes", msgLen, len(smData))
+	}
+
+	flags := binary.LittleEndian.Uint16(smData[9:11])
+	numSegments := binary.LittleEndian.Uint16(smData[11:13])
+	hasCRC := flags&SMFlagCRC64 != 0
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azdatalake/internal/shared/structured_message.go]: Same reasoning as above — the azdatalake decoder is a general-purpose SM parser. CRC64 enforcement happens at the transport level via header negotiation.

File: sdk/storage/azfile/internal/shared/structured_message.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azfile/internal/shared/structured_message.go:203]: Same reasoning as above — the azdatalake decoder is a general-purpose SM parser. CRC64 enforcement happens at the transport level via header negotiation.

File: sdk/storage/azfile/internal/shared/structured_message.go (hunk 18 above)
Comment (jhendrixMSFT) [reply sdk/storage/azfile/internal/shared/structured_message.go]: Thanks for clarifying.

Instead of relying on a specific error type (e.g. `net.Error`) it might be better to assume that _all_ errors are retriable.  For the errors you know _aren't_ retriable, wrap them with `/internal/errorinfo.NonRetriableError()`.  You can then use `errors.As` to check if an error is non-retriable (see [here](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azcore/runtime/policy_retry.go#L172-L178) in the retry policy for an example).

Early on in the retry policy we attempted to be more targeted for retries based on error type.  We quickly found out that it's just not reliable as there are too many error types and in some (rare) cases can be different across operating systems.  So, it's better to assume all errors are transient and annotate the ones that you know aren't.

Thoughts on this?

Comment (jhendrixMSFT) [reply sdk/storage/azfile/internal/shared/structured_message.go]: Also, I see some things in `retry_reader.go` that need to be retooled.  Might not be comprehensive but here are the ones that jump out at me.
```go
if err == nil || err == io.EOF // use errors.Is instead of comparison
...
_, isNetError := err.(net.Error) // use errors.As, or retool as above
...
const ReadOnClosedBodyMessage = "read on closed response body" // we shouldn't be exporting this
```

File: sdk/storage/azfile/internal/exported/transfer_validation_option.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (jhendrixMSFT) [reply sdk/storage/azfile/internal/exported/transfer_validation_option.go]: Was this resolved in https://github.com/Azure/azure-sdk-for-go/pull/27261#discussion_r3663842829?

File: sdk/storage/azfile/file/client.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azfile/file/client.go:625]: Fixed in 6124981.

Root cause: `SMDecoder.Read` deferred validation of the final segment footer and message trailer CRC64 to a *subsequent* `Read`. When the last payload read exactly filled the caller's buffer, the bounded `RetryReader` in the parallel download path (`DownloadBuffer`/`DownloadFile`) saw its raw range count reach zero and returned `io.EOF` before that follow-up `Read` ever happened — so the trailing CRC64s were never validated.

The decoder now drains and validates the trailing framing (final segment footer + message trailer) within the **same** `Read` call that emits the last payload byte. Framing bytes are read into an internal buffer rather than the caller's buffer, so this is independent of buffer alignment and validation always completes before EOF. Added `TestStreamingDecoderValidatesTrailerOnExactBufferFill` (plus a valid-path counterpart) covering the exact-buffer-fill boundary.

File: sdk/storage/azfile/internal/exported/transfer_validation_option.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azfile/internal/exported/transfer_validation_option.go]: Yes — the public `TransferValidationTypeCRC64` and `TransferValidationTypeComputeCRC64` exports were removed there (commit 8b1fc33).

In 6124981 I also removed the leftover **internal** precomputed/compute CRC64 options that delegated to the Azure Files `SetCRC64` no-op, and dropped the `reflect.Kind() == Func` branch from the upload guards. So there's no longer any `TransferValidationType` that can reach the no-op `SetCRC64`: only the structured-message path (`TransferValidationTypeComputeStructuredMessageCRC64`) remains, and any other non-structured-message transfer validation on the multi-chunk upload paths returns `UnsupportedChecksum`.

File: sdk/storage/azfile/internal/shared/structured_message.go (hunk 18 above)
Comment (tanyasethi-msft) [reply sdk/storage/azfile/internal/shared/structured_message.go]: Thanks. Addressed the concrete `retry_reader.go` items in 6124981:

- `err == nil || err == io.EOF` and the `io.ErrUnexpectedEOF` check now use `errors.Is`.
- The `net.Error` detection now uses `errors.As` instead of a direct type assertion. This one actually mattered for this PR: the decoder wraps segment-read failures with `%w`, and the old `err.(net.Error)` assertion did **not** unwrap the chain, so those wrapped `net.Error`s were silently *not* being retried. With `errors.As` the `%w` chain is honored and transient segment-read failures retry as intended.
- Unexported `ReadOnClosedBodyMessage` -> `readOnClosedBodyMessage` (it's only used within the package).

On the broader "assume every error is retriable and tag the known-non-retriable ones with `internal/errorinfo.NonRetriableError()`" model — I agree it's the more robust approach. Since this retry reader is currently a verbatim copy of the azblob and azdatalake ones, I'd rather make that conversion as a single focused change across all three storage retry readers so they stay in sync, rather than diverging azfile on its own. Happy to fold it into this PR instead if you'd prefer to keep it together.

File: sdk/storage/azfile/file/retry_reader.go [hunk 21]
```diff
@@ -165,11 +169,11 @@ func (s *RetryReader) wasRetryableEarlyClose(err error) bool {
...
 	}
 	// unfortunately, http.errReadOnClosedResBody is private, so the best we can do here is to check for its text
-	return strings.HasSuffix(err.Error(), ReadOnClosedBodyMessage)
+	return strings.HasSuffix(err.Error(), readOnClosedBodyMessage)
 }
 
-// ReadOnClosedBodyMessage of retry reader
-const ReadOnClosedBodyMessage = "read on closed response body"
+// readOnClosedBodyMessage of retry reader
+const readOnClosedBodyMessage = "read on closed response body"
```
Comment (jhendrixMSFT) [inline sdk/storage/azfile/file/retry_reader.go]: Turns out this has been in GA releases so will need to keep exporting it.

File: sdk/storage/azfile/internal/shared/structured_message.go (hunk 18 above)
Comment (jhendrixMSFT) [reply sdk/storage/azfile/internal/shared/structured_message.go]: Separate PR is fine and will help keep it targeted.

File: sdk/storage/azfile/file/retry_reader.go (hunk 21 above)
Comment (tanyasethi-msft) [reply sdk/storage/azfile/file/retry_reader.go]: Good catch — you're right, it's exported on `main` and shipped in GA, so unexporting it would be a breaking change. Reverted in 8f53360 — restored the exported `ReadOnClosedBodyMessage` constant and its reference in `wasRetryableEarlyClose`. Thanks!

File: sdk/storage/azfile/internal/shared/structured_message.go (hunk 18 above)
Comment (tanyasethi-msft) [reply sdk/storage/azfile/internal/shared/structured_message.go]: Sounds good — tracked in #27327. I'll do the `NonRetriableError()` conversion across all three storage retry readers (azblob/azdatalake/azfile) in that focused follow-up so they stay in sync. Thanks!

File: sdk/storage/azfile/internal/shared/structured_message.go [hunk 22]
```diff
@@ -0,0 +1,825 @@
...
+				// Segment data fully read — emit footer
+				footer := make([]byte, SMSegmentFooterSize)
+				binary.LittleEndian.PutUint64(footer, e.segCRC.Sum64())
+				e.pending = footer
+				e.pendingOff = 0
+				e.state = encStateSegFooter
+			}
+			if err != nil && e.segRemain > 0 {
+				return totalRead, err
+			}
```
[the remark this answers is not in the document]
Comment (jhendrixMSFT) [reply sdk/storage/azfile/internal/shared/structured_message.go]: Suggestion from my Copilot review.
```go

if err != nil {
    finalContentBoundary := e.segRemain == 0 && e.segIndex == e.numSegments
    if !errors.Is(err, io.EOF) || !finalContentBoundary {
        return totalRead, err
    }
}
```

[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azfile/internal/shared/structured_message.go]: Applied — thanks. `SMEncoder.Read` now only ignores `io.EOF` at the final content boundary (`segRemain == 0 && segIndex == numSegments`) and propagates every other error, so a source returning data together with a non-EOF error on a segment boundary no longer emits a valid trailer. Used `errors.Is(err, io.EOF)` as suggested.

I also switched the `err == io.EOF` check in `SMDecoder.fillFrame` to `errors.Is` for the same wrapped-error robustness, and added regression tests for both the non-EOF-error and data-with-EOF boundary cases (`TestEncoderPropagatesNonEOFErrorAtSegmentBoundary`, `TestEncoderAcceptsDataWithEOFAtFinalBoundary`).

**Review by jhendrixMSFT — APPROVED** [review]
